### PR TITLE
new SSO authentication flow

### PR DESCRIPTION
**Pull request by lgarvey**
This PR adds a new authentication flow that asks for the user's login email then directs them to the most appropriate identity provider.

### CODE REVIEW

File: sso/samlauth/views.py [hunk 1]
```diff
@@ -394,3 +405,72 @@ def logged_out(request):
...
+    def form_valid(self, form):
+
+        email = form.cleaned_data['email']
+
+        if form.idp_ref:
+            idp = self.lookup_idp_from_ref(form.idp_ref)
+            url = reverse('saml2_login') + f'?idp={idp}'
+            args = self.request.META.get('QUERY_STRING', '')
+
+            if args:
```
Comment (michalc) [inline sso/samlauth/views.py]: _Extremely_ minor point, but will it all work equivalently without this `if`, and doing the below line unconditionally?

File: sso/samlauth/forms.py [hunk 2]
```diff
@@ -0,0 +1,34 @@
+from django import forms
+from django.conf import settings
+
+
+def lookup_idp_ref_from_email(email_domain):
+    idp_ref = None
+    for ref, allowed_email_domains in settings.AUTH_EMAIL_TO_IPD_MAP.items():
+        if any(True for domain in allowed_email_domains if domain == email_domain):
```
Comment (michalc) [inline sso/samlauth/forms.py]: Can this be simplified to `if email_domain in allowed_email_domains:` ?

File: sso/samlauth/views.py [hunk 3]
```diff
@@ -394,3 +405,72 @@ def logged_out(request):
...
+
+        return [k for k, v in idps.items() if v == ref][0]
+
+    def form_valid(self, form):
+
+        email = form.cleaned_data['email']
+
+        if form.idp_ref:
+            idp = self.lookup_idp_from_ref(form.idp_ref)
+            url = reverse('saml2_login') + f'?idp={idp}'
```
Comment (michalc) [inline sso/samlauth/views.py]: If `idp` is an URL, to be robust about it containing things like `?` and `&`, should it be url-encoded?

File: sso/samlauth/views.py [hunk 4]
```diff
@@ -394,3 +405,72 @@ def logged_out(request):
...
+            try:
+                if User.objects.get(email=user_email).use_new_journey:
+                    choice = 'saml2_login_start'
+            except User.DoesNotExist:
+                pass
+
+        url = reverse(choice)
+
+        args = self.request.META.get('QUERY_STRING', '')
+        if args:
```
Comment (michalc) [inline sso/samlauth/views.py]: Similar extremely minor point to above

File: sso/tests/test_auth_flow.py [hunk 5]
```diff
@@ -715,3 +718,114 @@ def test_last_login_time_recorded_against_email(self, client, mocker):
...
+
+class TestEmailBasedAuthFlow:
+    def test_invalid_email_leads_to_form_error(self, client):
+        response = client.post(reverse('saml2_login_start'), {'email': 'test@invalid.com'})
+
+        assert response.status_code == 200
+
+        assert 'You cannot login with this email address'.encode('utf-8') in response.content
+
+    def test_previously_used_email_is_rendered_on_form(self, client):
```
Comment (michalc) [inline sso/tests/test_auth_flow.py]: A minor thing, but should this test be merged with `test_used_email_is_saved_in_cookie`, and make it so the tests check on what happens from the point of view of the user-facing behaviour of form inputs and what is shown in the forms, and not explicitly settings and checking cookies?

File: sso/samlauth/views.py (hunk 1 above)
Comment (lgarvey) [reply sso/samlauth/views.py]: Yep, the if only adds QA args if they exist.  But the url is completely formed before the if statement.

File: sso/samlauth/forms.py (hunk 2 above)
Comment (lgarvey) [reply sso/samlauth/forms.py]: I think it can - weird oversight.  I'll amend this.

File: sso/samlauth/views.py (hunk 3 above)
Comment (lgarvey) [reply sso/samlauth/views.py]: Good point - I'll add some url encoding. Interestingly I don't think the django RedirectView does this.

File: sso/tests/test_auth_flow.py (hunk 5 above)
Comment (lgarvey) [reply sso/tests/test_auth_flow.py]: I think these need to be kept separate as they happen at very different points of the auth flow.  It'd be useful to add some e2e tests that do broader tests at some point.

**Pull request merged**